File: benchmark/rclcpp/service/client-stress-test.cpp
// Copyright (c) 2018 Intel Corporation. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <chrono>
#include <iostream>
#include <memory>
#include <string>

#include "nav_msgs/srv/get_map.hpp"
#include "rclcpp/rclcpp.hpp"
#include "utilities.hpp"

void ShowUsage(const std::string name) {
  std::cerr << "Usage: " << name << " [options]\n"
            << "\nOptions:\n"
            << "\n-r, --run <n>     \tHow many times to run\n"
            << "-h, --help          \toutput usage information" << std::endl;
}

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);

  auto total_times = 0;
  for (int i = 1; i < argc; i++) {
    std::string arg = argv[i];
    if ((arg == "-h") || (arg == "--help")) {
      ShowUsage(argv[0]);
      return 0;
    } else if (arg.find("--run=") != std::string::npos) {
      total_times = std::stoi(arg.substr(arg.find("=") + 1));
    } else if (arg == "-r" && i + 1 < argc) {
      total_times = std::stoi(argv[++i]);
    }
  }
  printf(
      "The client will send a GetMap request continuously until receiving "
      "response %d times.\n",
      total_times);

  auto start = std::chrono::high_resolution_clock::now();
  auto node = rclcpp::Node::make_shared("stress_client_rclcpp");
  auto client = node->create_client<nav_msgs::srv::GetMap>("get_map");
  auto request = std::make_shared<nav_msgs::srv::GetMap::Request>();
  auto received_times = 0;

  // Wait for service to be available
  while (!client->wait_for_service(std::chrono::seconds(1))) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("rclcpp"),
                   "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"),
                "Service not available, waiting again...");
  }

  while (rclcpp::ok()) {
    if (received_times >= total_times) {
      rclcpp::shutdown();
      auto end = std::chrono::high_resolution_clock::now();
      LogTimeConsumption(start, end);
      break;
    } else {
      auto result_future = client->async_send_request(request);

      // Spin until future is ready using the node directly
      auto future_status = rclcpp::spin_until_future_complete(
          node, result_future, std::chrono::seconds(10));

      if (future_status != rclcpp::FutureReturnCode::SUCCESS) {
        RCLCPP_ERROR(node->get_logger(), "Service call failed or timed out.");
        return 1;
      }

      auto result = result_future.get();
      received_times++;
    }
  }
  return 0;
}

File: benchmark/rclpy/service/client-stress-test.py
#!/usr/bin/env python3
# Copyright (c) 2018 Intel Corporation. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import math
import time

import rclpy
from rclpy.node import Node
from nav_msgs.srv import GetMap

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("-r", "--run", type=int, default=1, help="How many times to run")
    args = parser.parse_args()

    rclpy.init()
    
    try:
        node = Node('stress_client_rclpy')
        
        # Create client
        client = node.create_client(GetMap, 'get_map')
        
        # Wait for service to be available
        node.get_logger().info('Waiting for service to be available...')
        if not client.wait_for_service(timeout_sec=5.0):
            node.get_logger().error('Service not available after waiting')
            return
        
        node.get_logger().info(f'The client will send a GetMap request continuously '
                              f'until receiving {args.run} response times.')
        
        start_time = time.time()
        received_times = 0
        
        # Send requests synchronously
        for i in range(args.run):
            request = GetMap.Request()
            try:
                future = client.call_async(request)
                rclpy.spin_until_future_complete(node, future, timeout_sec=5.0)
                
                if future.result() is not None:
                    received_times += 1
                else:
                    node.get_logger().error(f'Request {i+1} failed')
            except Exception as e:
                node.get_logger().error(f'Service call {i+1} failed: {e}')
        
        # Log completion
        diff = time.time() - start_time
        milliseconds, seconds = math.modf(diff)
        node.get_logger().info(f'Benchmark took {int(seconds)} seconds and '
                              f'{round(milliseconds * 1000)} milliseconds.')
        node.get_logger().info(f'Successfully received {received_times}/{args.run} responses')
        
        node.destroy_node()
    except KeyboardInterrupt:
        pass
    finally:
        rclpy.try_shutdown()

if __name__ == '__main__':
    main()
